### PR TITLE
Add flow-launcher:// deep link protocol, .flowplugin file association, and --query command line argument

### PR DESCRIPTION
**Pull request by Garulf**
## What

Adds a `flow-launcher://` deep link protocol and `.flowplugin` file association, plus a `--query "text"` (also `-query`) command-line flag, so scripts, browsers, and other apps can open/focus Flow Launcher to run a query, open settings, or install a plugin.

- **`--query "text"` / `-q "text"`**: opens/focuses Flow Launcher with the query pre-filled and searched.
- **`flow-launcher://query?q=...`**, **`flow-launcher://settings`**, **`flow-launcher://plugin/install?path=|id=|url=`**: routed through a general verb-dispatch framework. Plugin installs always show the existing Yes/No confirmation dialog — never silent, whether triggered by a local `.flowplugin` double-click, a store id, or an https download URL.
- **`.flowplugin` file association** and the **`flow-launcher://` URI scheme** are self-healing HKCU registrations applied at every startup (no admin rights required), with a new Settings toggle, "Allow websites to open Flow Launcher", to opt out.
- **Cold start** (no instance running): the deep link is dispatched right after plugin initialization, then the main window is shown.
- **Warm start** (instance already running): the second process normalizes its args to a `flow-launcher://` URI and sends it to the first instance over the existing single-instance named pipe, then exits; the first instance dispatches it.
- **No args**: behavior is unchanged — a second launch just shows/focuses the window.

## How

- `Flow.Launcher/Helper/DeepLink.cs`: normalizes command-line args (`--query`, `.flowplugin` paths, raw `flow-launcher://` URIs) into a single URI string, parses a URI into a verb + query parameters, and dispatches to handlers for `query`, `settings`, and `plugin/install`.
- `Flow.Launcher/Helper/DeepLinkRegistration.cs`: idempotent HKCU registration for the `.flowplugin` ProgID and the `flow-launcher://` scheme, called on every startup so stale executable paths self-heal after updates.
- `Flow.Launcher.Core/Plugin/PluginInstaller.cs`: adds `InstallPluginFromWebAndCheckRestartAsync(url)`, reusing the existing unknown-source warning and the always-prompting `InstallPluginAndCheckRestartAsync`.
- `App.xaml.cs`: `Main(string[] args)` normalizes args via `DeepLink.FromCommandLineArgs`; the pending deep link is dispatched after `PluginManager.InitializePluginsAsync`, before the home-page refresh check; `OnSecondAppStarted(string payload)` dispatches the warm-path payload (or just shows the window when empty).
- `SingleInstance.cs`: the named pipe now carries an optional payload — the normalized deep link URI (or legacy plain query text). The client writes it as a newline-terminated UTF-8 line and blocks (up to 3 s) since the process exits immediately after; the server reads one line guarded by a 2 s timeout and try/catch, so a client that connects but never writes degrades to a plain activation and can never hang the server loop.
- `Settings.EnableDeepLinkProtocol` (default `true`) gates the URI scheme registration; toggling it in Settings registers/unregisters the scheme immediately via `DeepLinkRegistration`.

## Test matrix (manual, Windows)

1. No args, cold start — unchanged behavior (home page etc.)
2. `--query "foo"` / `-query "foo"`, cold and warm — window opens/focuses with "foo" searched
3. Warm start, no args — window just shows/focuses (regression check)
4. `--query ""` — treated as no query
5. Double-click a `.flowplugin` (Flow running and not running) — confirmation dialog, install only on Yes
6. `flow-launcher://query?q=foo`, `flow-launcher://settings`, `flow-launcher://plugin/install?id=...` / `?path=...` / `?url=https://...zip` from a browser — routed correctly, every install prompts
7. `?url=http://...` (non-https) — rejected with an https-only message
8. `flow-launcher://nope` — error notification, no crash
9. Settings toggle off — `HKCU\Software\Classes\flow-launcher` removed, browser links no longer open Flow; `.flowplugin` double-click and `--query` still work. Toggle on — restored
10. Corrupt the registered command path manually, relaunch — self-healed
11. Robustness: a client that connects to the pipe without writing/closing — server loop recovers after the 2 s read timeout and keeps accepting connections

---
## Summary by cubic
**Summary of changes**
Adds a `flow-launcher://` deep link and `.flowplugin` association plus a `--query`/`-q` flag so scripts and websites can open Flow Launcher, run searches, open settings, or install plugins. When the URI scheme is disabled or no args are provided, launch behavior is unchanged; second-instance invocations forward a normalized deep link to the running app over IPC.

- Changed
  - Single-instance IPC now carries an optional UTF-8 deep-link payload; server cancels the read after 2s, client connect waits up to 3s, and the outer delivery wait is 5s so payloads are not dropped; dispatcher activation exceptions are observed.
  - Pipe read timeouts become plain activations; genuine read failures are logged.
  - `App.Main(string[] args)` normalizes args to a deep link; dispatch happens after plugins initialize. Links received during startup are queued so none are lost. `OnSecondAppStarted(string payload)` routes a payload; empty payload focuses the window.
  - CLI supports `--query` and `-q`; empty queries are ignored.
  - Deep link protocol toggle updates registry immediately and reverts the toggle if registration fails.
  - Logging keeps full deep-link payloads at Debug; Warn logs only the actionable fact.
- Added
  - Deep link parser/dispatcher with verbs: query, settings, and plugin/install (by `.flowplugin` path, store id, or https URL). Helper `HasExactlyOneInstallIdentifier` validates install links.
  - HKCU registration for `.flowplugin` and the `flow-launcher://` scheme, gated by a new General Settings toggle “Allow websites to open Flow Launcher”; registration is ensured at startup.
  - `PluginInstaller.InstallPluginFromWebAndCheckRestartAsync` with unknown-source confirmation; `DerivePluginNameFromUrl` sanitizes filenames and falls back to a safe default; `TryPopulateIdFromZip` recovers the real plugin ID for URL installs to keep modified-plugin tracking correct.
  - UI strings and a General Settings toggle. Unit tests: `DeepLinkTest` (arg normalization, scheme-disabled behavior, URI parsing) and `PluginInstallerTest` (URL filename derivation edge cases).
- Removed
  - Old `OnSecondAppStarted()` overload; no user-visible features removed.
- Memory
  - Negligible: short-lived strings and small stream buffers; no persistent allocations.
- Security
  - URI scheme is opt-in and per-user (HKCU). Named pipe remains per-user. Deep links are parsed/validated; plugin installs require https URLs or an existing local file; store-id installs validate against the manifest. Payload logging is minimized; IPC timeouts prevent hangs.
- Tests
  - Added `DeepLinkTest` and `PluginInstallerTest`; manual verification of cold/warm starts, empty queries, startup-queued links, and all plugin install paths.

**Release Note**
Open Flow Launcher from your browser or scripts to search, open settings, or install plugins using special links or the new --query option.

<sup>Written for commit cd2a1b91899dd1374ab320852bd492a4b22d18ce. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/Flow-Launcher/Flow.Launcher/pull/4565?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: Flow.Launcher/App.xaml.cs [hunk 1]
```diff
@@ -151,14 +152,28 @@ public static void Main()
...
                 application.InitializeComponent();
                 application.Run();
             }
         }
 
+        private static string ParseQueryArg(string[] args)
+        {
+            for (var i = 0; i < args.Length; i++)
+            {
+                if ((args[i] == "--query" || args[i] == "-query") && i + 1 < args.Length)
```
Comment (DavidGBrett) [inline Flow.Launcher/App.xaml.cs]: Shouldn't the single dash argument be a single character?

i.e `-q` instead of `-query`

Comment (Garulf) [reply Flow.Launcher/App.xaml.cs]: Makes sense to me

**Comment by Garulf** [comment]
Personally tested:
- [x] Plugin install URI by ID `flow-launcher://plugin/install?id=B2D2C23B055D411GH66FF0C79D6C2A6H`
- [x] Plugin install URI by URL `flow-launcher://plugin/install?url=https://github.com/Garulf/HA-Commander/releases/download/v5.2.1/HA-Commander.zip`
- [x] Plugin install URI by path `flow-launcher://plugin/install?path=D:\Users\garulf\Downloads\HA-Commander.zip` 
- [x] Query URI `flow-launcher://query?q=helloworld`
- [x] Settings URI `flow-launcher://settings`
- [x] CLI args `./Flow.Launcher.exe -q "hi"`
- [x] Install plugin zip with `.flowplugin` extension
- [x] Disabling "Allow websites to open Flow Launcher" stops URI from working
- [x] Enabling"Allow websites to open Flow Launcher" allows URI to work

Everything has been tested to work

**Comment by Jack251970** [comment]
@codex Review it

File: Flow.Launcher.Core/Plugin/PluginInstaller.cs [hunk 2]
```diff
@@ -148,6 +148,51 @@ public static async Task InstallPluginAndCheckRestartAsync(string filePath)
         await InstallPluginAndCheckRestartAsync(plugin);
     }
 
+    /// <summary>
+    /// Installs a plugin from a direct zip download URL and restarts the application if required by settings.
+    /// Applies the unknown source warning and prompts user for confirmation before installing.
+    /// </summary>
+    /// <param name="url">The https URL of the plugin zip file.</param>
+    /// <returns>A Task representing the asynchronous install operation.</returns>
+    public static async Task InstallPluginFromWebAndCheckRestartAsync(string url)
```
Comment (VictoriousRaptor) [inline Flow.Launcher.Core/Plugin/PluginInstaller.cs]: We currently do support `pm install https://some.plugin.zip`. Is this function any different?

File: Flow.Launcher/Helper/DeepLinkRegistration.cs [hunk 3]
```diff
@@ -0,0 +1,76 @@
+using System;
+using Flow.Launcher.Infrastructure;
+using Microsoft.Win32;
+
+namespace Flow.Launcher.Helper;
+
+/// <summary>
+/// Registers the .flowplugin file association and the flow-launcher:// URI scheme
+/// under HKCU\Software\Classes. Writes are idempotent, so calling on every startup
```
Comment (VictoriousRaptor) [inline Flow.Launcher/Helper/DeepLinkRegistration.cs]: Can uninstaller cleans this registry?

File: Flow.Launcher/Helper/DeepLink.cs [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+    /// <summary>
+    /// Normalizes command line arguments to a single deep link URI string, or null for a normal launch.
+    /// Raw flow-launcher:// arguments are dropped when <paramref name="allowSchemeArgs"/> is false
+    /// (URI scheme disabled in settings); --query/-q and .flowplugin paths are always honored.
+    /// </summary>
+    public static string FromCommandLineArgs(string[] args, bool allowSchemeArgs)
+    {
+        for (var i = 0; i < args.Length; i++)
+        {
+            if (args[i] == "--query" || args[i] == "-q")
```
[the remark this answers is not in the document]
Comment (Garulf) [reply Flow.Launcher/Helper/DeepLink.cs:41]: `-query` never shipped, it was introduced in this PR and renamed to `-q` at a reviewer's request before merge, so there is no backward compatibility to preserve. The test matrix in the PR comments already lists `-q "hi"`.

File: Flow.Launcher.Core/Plugin/PluginInstaller.cs [hunk 5]
```diff
@@ -148,6 +148,51 @@ public static async Task InstallPluginAndCheckRestartAsync(string filePath)
...
+    {
+        // Derive the filename from the URI path so query strings (e.g. plugin.zip?token=x) don't end up
+        // as part of a temp filename, which is invalid on Windows. Fall back to the naive split if the
+        // URL somehow fails to parse as a URI (callers already validate https, this is defense in depth).
+        // uri.LocalPath is percent-decoded, so it can still contain characters that are invalid in
+        // Windows filenames (e.g. a literal "?" from an encoded "%3F"); strip those out.
+        var filename = Uri.TryCreate(url, UriKind.Absolute, out var uri)
+            ? Path.GetFileName(uri.LocalPath)
+            : url.Split('/').Last();
+        foreach (var c in Path.GetInvalidFileNameChars())
```
[the remark this answers is not in the document]
Comment (Garulf) [reply Flow.Launcher.Core/Plugin/PluginInstaller.cs]: Fixed in 568b3f845: when sanitization collapses the basename to an empty string the name now falls back to `plugin`, with unit tests covering the encoded-invalid-char and no-basename cases.

File: Flow.Launcher/App.xaml.cs [hunk 6]
```diff
@@ -465,9 +477,15 @@ public void Dispose()
...
+        public void OnSecondAppStarted(string payload)
         {
-            API.ShowMainWindow();
+            if (string.IsNullOrEmpty(payload))
+            {
+                API.ShowMainWindow();
+                return;
+            }
+
+            DeepLink.Dispatch(payload);
```
[the remark this answers is not in the document]
Comment (Garulf) [reply Flow.Launcher/App.xaml.cs:519]: Fixed in 568b3f845: `OnSecondAppStarted` now queues the payload into `_pendingDeepLink` (behind a lock shared with the startup flush) until plugin initialization completes, and the startup path dispatches whatever is pending at that point, latest link wins.

File: Flow.Launcher.Core/Plugin/PluginInstaller.cs [hunk 7]
```diff
@@ -148,6 +148,51 @@ public static async Task InstallPluginAndCheckRestartAsync(string filePath)
...
+        {
+            filename = filename.Replace(c.ToString(), string.Empty);
+        }
+        var name = filename.EndsWith(".zip", StringComparison.OrdinalIgnoreCase)
+            ? filename[..^".zip".Length]
+            : filename;
+
+        var plugin = new UserPlugin
+        {
+            ID = string.Empty,
```
[the remark this answers is not in the document]
Comment (Garulf) [reply Flow.Launcher.Core/Plugin/PluginInstaller.cs:177]: Fixed in 568b3f845: after the download completes, the real ID is read from the zip's plugin.json before entering the install path, so `ModifiedPlugins` tracks the actual plugin ID and later URL installs are no longer rejected. If the zip is unreadable the ID stays empty and the existing `InstallPlugin` error path reports it.

File: Flow.Launcher.Core/Plugin/PluginInstaller.cs [hunk 8]
```diff
@@ -151,6 +161,86 @@ public static async Task InstallPluginAndCheckRestartAsync(string filePath)
...
+        {
+            using var archive = ZipFile.OpenRead(zipFilePath);
+            var pluginJsonEntry = archive.Entries.FirstOrDefault(x => x.Name == "plugin.json");
+            if (pluginJsonEntry == null) return;
+
+            using var stream = pluginJsonEntry.Open();
+            var zipPlugin = JsonSerializer.Deserialize<UserPlugin>(stream);
+            if (!string.IsNullOrEmpty(zipPlugin?.ID))
+            {
+                plugin.ID = zipPlugin.ID;
```
[the remark this answers is not in the document]
Comment (Garulf) [reply Flow.Launcher.Core/Plugin/PluginInstaller.cs:233]: This is parity with existing behavior rather than new attack surface: the local `.flowplugin` install path (`InstallPluginAndCheckRestartAsync(string)`) already deserializes `plugin.json` from an arbitrary zip and passes its ID through the same routing in `PluginManager.InstallPlugin`, and that path is reachable from the same deep link via `?path=`. URL installs are additionally gated by the unknown-source warning and the explicit install prompt. Hardening the default-ID routing against ID spoofing would need to happen in `PluginManager.InstallPlugin` for all install sources, so it is better tracked as a separate issue than a URL-only special case here.
